pkg/network/containers/container_store_linux.go

```diff
@@ -0,0 +1,236 @@
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"go4.org/intern"
+
+	"github.com/DataDog/datadog-agent/pkg/network"
+	"github.com/DataDog/datadog-agent/pkg/network/events"
+	"github.com/DataDog/datadog-agent/pkg/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// containerItemNoDataError is returned when readContainerItem() didn't fail, but doesn't
+// have data to share. This can happen when the process exited at the same time, or on unsupported platforms
+type containerItemNoDataError struct {
+	Err error
+}
+
+func (e *containerItemNoDataError) Error() string {
+	return fmt.Sprintf("readContainerItem() doesn't have data: %s", e.Err)
+}
+func (e *containerItemNoDataError) Unwrap() error {
+	return e.Err
+}
+
+const (
+	resolvConfMaxSizeBytes = 1024
+	maxProcessQueueLen     = 100
+	moduleName             = "network_tracer__containerStore"
+
+	// containerStaleTime: when to re-fetch a container's resolv.conf
+	containerStaleTime = 2 * time.Minute
+	// containerTTL: when to evict a container
+	containerTTL = 5 * time.Minute
+	// cleanerInterval: how often to evict old containers
+	cleanerInterval = 2 * time.Minute
+)
+
+var containerStoreTelemetry = struct {
+	liveEvictions telemetry.Counter
+	eventsDropped telemetry.Counter
+	readFailures  telemetry.Counter
+}{
+	telemetry.NewCounter(moduleName, "live_evicts", []string{}, "Counter measuring the number of evictions of live containers in the container store (this is bad)"),
+	telemetry.NewCounter(moduleName, "events_dropped", []string{}, "Counter measuring the number of dropped process events"),
+	telemetry.NewCounter(moduleName, "read_failures", []string{}, "Counter measuring the number of failures to read container data such as resolv.conf"),
+}
+
+type containerStoreItem struct {
+	timestamp  time.Time
+	resolvConf network.ResolvConf
+}
+
+// ContainerStore reads container data (currently just resolv.conf) into a hashmap
+// which is later attached to connections via the ResolvConf field
+type ContainerStore struct {
+	ctx       context.Context
+	cancelCtx func()
+
+	cache *lru.Cache[network.ContainerID, containerStoreItem]
+
+	in chan *events.Process
+
+	warnLimit  *log.Limit
+	errorLimit *log.Limit
+
+	readContainerItem func(ctx context.Context, entry *events.Process) (containerStoreItem, error)
+}
+
+func (csi containerStoreItem) isStale() bool {
+	return time.Since(csi.timestamp) > containerStaleTime
+}
+func (csi containerStoreItem) isExpired() bool {
+	return time.Since(csi.timestamp) > containerTTL
+}
+
+// NewContainerStore initializes the container store
+func NewContainerStore(maxContainers int) (*ContainerStore, error) {
+	warnLimit := log.NewLogLimit(5, 10*time.Second)
+	errorLimit := log.NewLogLimit(5, 10*time.Second)
+
+	cache, err := lru.NewWithEvict(maxContainers, func(_key *intern.Value, item containerStoreItem) {
+		if !item.isStale() {
+			log.Errorf("CNM ContainerStore has more than %d live containers, was forced to evict a live container", maxContainers)
```

Comment:
My gut says this is more of a warning than an error. Nothing actually went wrong, there are just a lot of containers. WDYT?